manticore/core/smtlib/expression.py

```diff
@@ -159,105 +238,89 @@ def __rxor__(self, other):
-        self._value = value
-        super().__init__(*args, **kwargs)
+class BoolConstant(Bool, Constant):
+    def __init__(self, value: bool, **kwargs):
+        super().__init__(value=bool(value), **kwargs)
 
     def __bool__(self):
-        return self.value
-
-    @property
-    def value(self):
         return self._value
 
+    def __eq__(self, other):
+        # A class that overrides __eq__() and does not define __hash__()
+        # will have its __hash__() implicitly set to None.
+        if self.taint:
+            return super().__eq__(other)
+        return self.value == other
 
-class BoolOperation(Bool):
-    __slots__ = ["_operands"]
+    def __hash__(self):
+        return object.__hash__(self)
 
-    def __init__(self, *operands, **kwargs):
-        self._operands = operands
 
-        # If taint was not forced by a keyword argument, calculate default
-        kwargs.setdefault("taint", reduce(lambda x, y: x.union(y.taint), operands, frozenset()))
+class BoolOperation(Bool, Operation, abstract=True):
+    """ It's an operation that results in a Bool """
 
-        super().__init__(**kwargs)
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-    @property
-    def operands(self):
-        return self._operands
+    def __bool__(self):
+        # FIXME: TODO: re-think is we want to be this forgiving every use of
+        #  local_simplify looks hacky
+        simplified = local_simplify(self)
+        if isinstance(simplified, Constant):
+            return simplified.value
+        raise ExpressionError("BoolOperation can not be reduced to a constant")
 
 
 class BoolNot(BoolOperation):
-    def __init__(self, value, **kwargs):
-        super().__init__(value, **kwargs)
+    def __init__(self, operand: Bool, **kwargs):
+        super().__init__(operands=(operand,), **kwargs)
 
 
 class BoolAnd(BoolOperation):
-    def __init__(self, a, b, **kwargs):
-        super().__init__(a, b, **kwargs)
+    def __init__(self, operanda: Bool, operandb: Bool, **kwargs):
+        super().__init__(operands=(operanda, operandb), **kwargs)
 
 
 class BoolOr(BoolOperation):
-    def __init__(self, a: "Bool", b: "Bool", **kwargs):
-        super().__init__(a, b, **kwargs)
+    def __init__(self, operanda: Bool, operandb: Bool, **kwargs):
+        super().__init__(operands=(operanda, operandb), **kwargs)
 
 
 class BoolXor(BoolOperation):
-    def __init__(self, a, b, **kwargs):
-        super().__init__(a, b, **kwargs)
+    def __init__(self, operanda: Bool, operandb: Bool, **kwargs):
+        super().__init__(operands=(operanda, operandb), **kwargs)
 
 
 class BoolITE(BoolOperation):
-    def __init__(self, cond: "Bool", true: "Bool", false: "Bool", **kwargs):
-        super().__init__(cond, true, false, **kwargs)
+    def __init__(self, cond: Bool, true: Bool, false: Bool, **kwargs):
+        super().__init__(operands=(cond, true, false), **kwargs)
+
 
+class Bitvec(Expression, abstract=True):
```

Comment:
I'm torn on this. On the one hand, "Bitvec" feels better than "BitVec" to me since we rarely (if ever) actually refer to it as the expanded "Bit Vector." On the other, doing a global find-and-replace from "Bitvec" to "BitVec" reduces the size of this diff by about 20%, and the name change may break external scripts that use the BitVec types directly. Yes, they should probably use the ConstraintSet helper methods and Operators instead, but I know I've done it a few times, so others may have too. 

I'm leaning toward the latter. Thoughts?

Reply:
Thank you for diving into this MPR!
I do not have too much stake in the naming part but I think we should consider making all the arguments of the constructor of the multi-inheritance expression system to have _only_ keyword arguments. As prescribed in this old blog post: https://rhettinger.wordpress.com/2011/05/26/super-considered-super/.

If we go that way eventually all uses of those will be changed.

Reply:
>  I think we should consider making all the arguments of the constructor of the multi-inheritance expression system to have only keyword arguments.

I agree with this 👍

Reply:
BTW currently (in this PR) only the basic types Bool, Bitvec, Expression, Operation, Constant... force you to use keyword args. But the user facing types (the leafs in the inheritance tree ) still use positional args. Mostly to keep the diff "smaller".